### PR TITLE
Enable `--check-cfg` by default in UI tests

### PR DESCRIPTION
**Pull request by Urgau**
This PR enables-by-default `--check-cfg` in UI tests, now that it has become stable.

To do so this PR does 2 main things:
 - it introduce the `no-auto-check-cfg` directive to `compiletest`, to prevent any `--check-cfg` args (only to be used for `--check-cfg` tests)
 - it updates the _remaining_[^1] UI tests by either:
     - allowing the lint when neither expecting the lint nor giving the check-cfg args make sense
     - give the appropriate check-cfg args
     - or expect the lint, when it useful

[^1]: some preparation work was done in #123577 #123702

I highly recommend reviewing this PR commit-by-commit.

r? @jieyouxu

### CODE REVIEW

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -1027,12 +1027,31 @@ impl<'test> TestCx<'test> {
     }
 
     fn set_revision_flags(&self, cmd: &mut Command) {
+        // Normalize revisions to be lowercase and replace `-`s with `_`s.
+        // Otherwise the `--cfg` flag is not valid.
+        let normalize_revision = |revision: &str| revision.to_lowercase().replace("-", "_");
```
Comment (jieyouxu) [inline src/tools/compiletest/src/runtest.rs:1032]: Wait, what happens if a test writer specifies both `foo-bar` and `foo_bar` (an edge case that's unlikely to be hit, but the less jank the better)? Do we try to emit `--cfg foo_bar` twice?

File: src/tools/compiletest/src/runtest.rs [hunk 2]
```diff
@@ -1027,12 +1027,31 @@ impl<'test> TestCx<'test> {
...
         }
+
+        if !self.props.no_auto_check_cfg {
+            let mut check_cfg = String::with_capacity(25);
+
+            // Generate `cfg(FALSE, REV1, ..., REVN)` (for all possible revisions)
+            //
+            // For compatibility reason we consider the `FALSE` cfg to be expected
+            // since it is extensively used in the testsuite.
+            check_cfg.push_str("cfg(FALSE");
```
Comment (jieyouxu) [inline src/tools/compiletest/src/runtest.rs:1046]: Do we have any documentation for this special case? If not, could you please document this behavior in the dev-guide so that the knowledge of `FALSE` being "special" exists *somewhere*?

File: src/tools/compiletest/src/header.rs [hunk 3]
```diff
@@ -249,6 +251,7 @@ mod directives {
     pub const COMPARE_OUTPUT_LINES_BY_SUBSET: &'static str = "compare-output-lines-by-subset";
     pub const LLVM_COV_FLAGS: &'static str = "llvm-cov-flags";
     pub const FILECHECK_FLAGS: &'static str = "filecheck-flags";
+    pub const NO_AUTO_CHECK_CFG: &'static str = "no-auto-check-cfg";
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header.rs:254]: Could you also please document what this directive does in the dev-guide? I.e. when/why is this needed?

File: src/tools/compiletest/src/header.rs [hunk 4]
```diff
@@ -208,6 +208,8 @@ pub struct TestProps {
     pub llvm_cov_flags: Vec<String>,
     /// Extra flags to pass to LLVM's `filecheck` tool, in tests that use it.
     pub filecheck_flags: Vec<String>,
+    /// Don't automaticly insert any `--check-cfg` args
```
Comment (jieyouxu) [inline src/tools/compiletest/src/header.rs:211]:
```suggestion
    /// Don't automatically insert any `--check-cfg` args
```

:3

**Review by jieyouxu — APPROVED** [review]
The code changes generally look good to me. Just one question about revision normalization and some doc requests.

File: src/tools/compiletest/src/runtest.rs (hunk 1 above)
Comment (Urgau) [reply src/tools/compiletest/src/runtest.rs:1032]: Pre-existing, but yes I think so.

File: src/tools/compiletest/src/header.rs (hunk 3 above)
Comment (Urgau) [reply src/tools/compiletest/src/header.rs:254]: The documentation will need to be very conservative, but sure I will do it together with `FALSE`.

File: src/tools/compiletest/src/runtest.rs (hunk 2 above)
Comment (Urgau) [reply src/tools/compiletest/src/runtest.rs:1046]: I'm not aware, nor was I able to find any documentation regarding `FALSE` being a de-facto cfg that will eval to false.

> could you please document this behavior in the dev-guide

Sure, I will send for that.

File: src/tools/compiletest/src/runtest.rs (hunk 1 above)
Comment (Urgau) [reply src/tools/compiletest/src/runtest.rs:1032]: The compiler does handle duplicate `--cfg` arguments fine; but that would never be the case here, only one `--cfg` argument is set per revision. The only issue with similar names is that inside the code it would be impossible to distinguish between the two revisions.

File: src/tools/compiletest/src/header.rs (hunk 3 above)
Comment (jieyouxu) [reply src/tools/compiletest/src/header.rs:254]: That's fine yeah

File: src/tools/compiletest/src/runtest.rs (hunk 2 above)
Comment (Urgau) [reply src/tools/compiletest/src/runtest.rs:1046]: rustc-dev-guide PR is up https://github.com/rust-lang/rustc-dev-guide/pull/1966

**Comment by jieyouxu** [comment]
Thanks!

@bors r+ rollup

**Comment by matthiaskrgr** [comment]
@bors r-
https://github.com/rust-lang/rust/pull/124358#issuecomment-2076280677

**Comment by Urgau** [comment]
> [#124358 (comment)](https://github.com/rust-lang/rust/pull/124358#issuecomment-2076280677)

Hum, apparently the `ui-fulldeps` test suite in run with stage0 which doesn't yet know that `--check-cfg` has been stablized.

Fortunately the branching of master should happen tomorrow and the update of stage0 sometime next week, so I think the best course of action is just to wait a couple days.

**Comment by Urgau** [comment]
stage0 was updated in #124521, I rebase over it and tested it locally

@rustbot labels -S-blocked +S-waiting-on-review

**Review by jieyouxu — APPROVED** [review]

**Comment by Urgau** [comment]
> Failure in rollup: [#124628 (comment)](https://github.com/rust-lang/rust/pull/124628#issuecomment-2091132297) `tests/ui/sse2.rs` needs to be updated (recently modified, namely in #124597).

Raced by less than 2 hours :laughing: 

----

Fixed the test.

@rustbot ready

**Comment by fmease** [comment]
Sorry this has failed in yet another rollup: https://github.com/rust-lang/rust/pull/124633#issuecomment-2091895751

@bors r-

**Comment by Urgau** [comment]
> Sorry this has failed in yet another rollup: [#124633 (comment)](https://github.com/rust-lang/rust/pull/124633#issuecomment-2091895751)

No worries. The issue was in the `tests/ui/instrument-coverage/on-values.rs` test which is ignored by default (requires profiler support). Fixed the test, CI passes, let's try again.

@bors r=@jieyouxu

**Comment by Urgau** [comment]
Issue with two other tests, they were again ignored by-default, fixed them; and did a grep against the repo for similar issue and didn't find anything else. CI passes, let's try again.

@bors r=@jieyouxu
